airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java

```diff
@@ -105,53 +106,83 @@ private void runInternal(final IntegrationConfig parsed) throws Exception {
+          validateConfig(integration.spec().getConnectionSpecification(), config, "DISCOVER");
+          outputRecordCollector.accept(new AirbyteMessage().withType(Type.CATALOG).withCatalog(source.discover(config)));
+        }
+        // todo (cgardens) - it is incongruous that that read and write return airbyte message (the
+        // envelope) while the other commands return what goes inside it.
+        case READ -> {
+          final JsonNode config = parseConfig(parsed.getConfigPath());
+          validateConfig(integration.spec().getConnectionSpecification(), config, "READ");
+          final ConfiguredAirbyteCatalog catalog = parseConfig(parsed.getCatalogPath(), ConfiguredAirbyteCatalog.class);
+          final Optional<JsonNode> stateOptional = parsed.getStatePath().map(IntegrationRunner::parseConfig);
+          try (final AutoCloseableIterator<AirbyteMessage> messageIterator = source.read(config, catalog, stateOptional.orElse(null))) {
+            produceMessages(messageIterator);
+          }
+        }
+        // destination only
+        case WRITE -> {
+          final JsonNode config = parseConfig(parsed.getConfigPath());
+          validateConfig(integration.spec().getConnectionSpecification(), config, "WRITE");
+          final ConfiguredAirbyteCatalog catalog = parseConfig(parsed.getCatalogPath(), ConfiguredAirbyteCatalog.class);
+          try (final AirbyteMessageConsumer consumer = destination.getConsumer(config, catalog, outputRecordCollector)) {
+            runConsumer(consumer);
+          }
         }
+        default -> throw new IllegalStateException("Unexpected value: " + parsed.getCommand());
       }
-      default -> throw new IllegalStateException("Unexpected value: " + parsed.getCommand());
+    } catch (final Exception e) {
+      final String displayMessage = getDisplayMessage(e);
+      // If the source connector throws a config error, a trace message with the relevant message should
+      // be surfaced.
+      if (isConfigError(e)) {
+        AirbyteTraceMessageUtility.emitConfigErrorTrace(e, displayMessage);
+      }
+      if (parsed.getCommand().equals(Command.CHECK)) {
+        // Currently, special handling is required for the SPEC case since the user display information in
+        // the trace message is
+        // not properly surfaced to the FE. In the future, we can remove this and just throw an exception.
+        outputRecordCollector.accept(new AirbyteMessage().withType(Type.CONNECTION_STATUS).withConnectionStatus(
+            new AirbyteConnectionStatus().withStatus(AirbyteConnectionStatus.Status.FAILED).withMessage(displayMessage)));
+        return;
+      }
+      throw e;
     }
 
     LOGGER.info("Completed integration: {}", integration.getClass().getName());
   }
 
+  private boolean isConfigError(final Exception e) {
+    return e instanceof ConfigErrorException;
+  }
+
+  private String getDisplayMessage(final Exception e) {
+    if (e instanceof ConfigErrorException) {
```

Comment:
suggestion if this `getDisplayMessage` only applies to `ConfigErrorException` then you can probably have this as a ternary but not blocking or necessary

```java
return (e instanceof ConfigErrorException) ? ((ConfigErrorException) e).getDisplayMessage() : String.format("Could not connect with provided configuration. Error: %s", e.getMessage());
```